### PR TITLE
[ISSUE-2572] Set per scan node dop

### LINKED ISSUE #2572: ScanNode specific dop is set to speedup disk io
## Enhancement

For ScanOperator in pipeline engine, the number of concurrent io tasks is equivalant to the number of ScanOperator instances in a pipeline, the number of tablets, rows and fragment instances detemine how many rows will be fetched by ScanOperaors in a fragment instance. Different ScanOperators have different number of tablets, if the the number of tablets of some ScanOperator is quite large, then the ScanOperator should be assigned a large ScanNode-specific dop.

FE computes degree of parallelism(dop) for each OlapScanNode in a fragment instance. the maxScanOperators is numRows of ScanNode divided by session variable pipelineMaxRowCountPerScanOperator, then each fragment instance take the share of itself own numScanRanges divided by totalNumScanRanges, and the dop of OlapScanNode is at least 1 and at most minimum(itself own numScanRanges, half the avg number of be cores).

If a operator following a ScanOperator directly has a different dop with the ScanOperator, then a Passthrougth operator is interpolated in the middle.

### PR DESCRIPTION
**Pull request by satanson**
## Enhancement

For ScanOperator in pipeline engine, the number of concurrent io tasks is equivalant to the number of ScanOperator instances in a pipeline, the number of tablets, rows and fragment instances detemine how many rows will be fetched by ScanOperaors in a fragment instance. Different ScanOperators have different number of tablets, if the the number of tablets of some ScanOperator is quite large, then the ScanOperator should be assigned a large ScanNode-specific dop.

FE computes degree of parallelism(dop) for each OlapScanNode in a fragment instance. the maxScanOperators is numRows of ScanNode divided by session variable pipelineMaxRowCountPerScanOperator, then each fragment instance take the share of itself own numScanRanges divided by totalNumScanRanges, and the dop of OlapScanNode is at least 1 and at most minimum(itself own numScanRanges, half the avg number of be cores).

If a operator following a ScanOperator directly has a different dop with the ScanOperator, then a Passthrougth operator is interpolated in the middle.

### CODE REVIEW

File: be/src/exec/pipeline/pipeline_builder.h [hunk 1]
```diff
@@ -55,13 +58,22 @@ class PipelineBuilderContext {
...
+    }
+
 private:
     FragmentContext* _fragment_context;
     Pipelines _pipelines;
     uint32_t _next_pipeline_id = 0;
     uint32_t _next_operator_id = 0;
     int32_t _next_pseudo_plan_node_id = Operator::s_pseudo_plan_node_id_upper_bound;
     size_t _degree_of_parallelism = 1;
+    std::map<TPlanNodeId, int> _per_scan_node_dop;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_builder.h:76]: unordered_map

File: be/src/exec/pipeline/pipeline_builder.h [hunk 2]
```diff
@@ -55,13 +58,22 @@ class PipelineBuilderContext {
 
     FragmentContext* fragment_context() { return _fragment_context; }
 
+    size_t get_dop_of_scan_node(TPlanNodeId id) const {
+        if (_per_scan_node_dop.count(id)) {
+            return _per_scan_node_dop.at(id);
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_builder.h:63]:
```suggestion
            return _per_scan_node_dop[id];
```

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java [hunk 3]
```diff
@@ -188,6 +194,22 @@ public ExchangeNode getDestNode() {
         return destNode;
     }
 
+    public List<PlanNode> getOlapScanNodes(PlanNode root) {
+        if (root instanceof OlapScanNode) {
+            return Lists.newArrayList(root);
+        }
+
+        if (root.getChildren().isEmpty()) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:202]: The  List<PlanNode> scanNodes could pass by args, then we needn't this if branch, and could reduce list object generate.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 4]
```diff
@@ -309,6 +311,9 @@
     @VariableMgr.VarAttr(name = PIPELINE_DOP)
     private int pipelineDop = 0;
 
+    @VariableMgr.VarAttr(name = PIPELINE_MAX_ROW_COUNT_PER_SCAN_OPERATOR)
+    private long pipelineMaxRowCountPerScanOperator = 262144;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:315]: This variable could be invisible for users. `flag = VariableMgr.INVISIBLE`.

Why max row count per san operator so small?

File: be/src/exec/pipeline/pipeline_builder.h (hunk 2 above)
Comment (satanson) [reply be/src/exec/pipeline/pipeline_builder.h:63]: get_dop_of_scan_node is a const-modified function, operator[] is not const, instead , at(size_t i) is its const counterpart.

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java (hunk 3 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:202]: ok

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 4 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:315]: this variable should be assigned an robust and  emprical value, here the value is quite heuristic(that is to say, inaccurate), a better suggestion is welcome.

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 5]
```diff
@@ -1271,7 +1314,27 @@ private void computeFragmentHosts() throws Exception {
...
+                    OlapScanNode scanNode = (OlapScanNode) leftMostNode;
+                    int numScanRangesPerInstance = scanNode.getScanRangeLocations(0).size() / numInstances;
+                    numScanRangesPerInstance = Math.max(1, numScanRangesPerInstance);
+                    long numRows = scanNode.getCardinality();
+                    if (scanNode.hasLimit()) {
+                        numRows = Math.min(numRows, scanNode.getLimit());
+                    }
+
+                    float inputBytes = numRows * scanNode.getAvgRowSize();
+                    long maxBytesPerOperator = ConnectContext.get().getSessionVariable().getPipelineMaxInputBytesPerOperator();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1326]: Why here use input_bytes here? not row count?

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 6]
```diff
@@ -913,6 +915,47 @@ private void computeFragmentExecParams() throws Exception {
...
+        for (FragmentExecParams params : fragmentExecParamsMap.values()) {
+            List<PlanNode> scanNodes = params.fragment.getOlapScanNodes(params.fragment.getPlanRoot());
+            for (PlanNode node : scanNodes) {
+                OlapScanNode scanNode = (OlapScanNode) node;
+                int totalNumScanRanges = scanNode.getScanRangeLocations(0).size();
+                long numRows = scanNode.getCardinality();
+                if (scanNode.hasLimit()) {
+                    numRows = Math.min(scanNode.getLimit(), numRows);
+                }
+                int maxNumScanOperators = Math.max(1, (int) (numRows / maxRowCountPerScanOperator));
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:938]: maxNumScanOperators not used.

Why not consider input bytes here?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 4 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:318]: This variable could be invisible for users. flag = VariableMgr.INVISIBLE. We shouldn't  expect the user change the value.

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java [hunk 7]
```diff
@@ -194,6 +194,18 @@ public ExchangeNode getDestNode() {
         return destNode;
     }
 
+    public List<PlanNode> getOlapScanNodes(PlanNode root) {
+        if (root instanceof OlapScanNode) {
+            return Lists.newArrayList(root);
+        }
+
+        List<PlanNode> scanNodes = Lists.newArrayList();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:202]: The  `List<PlanNode> scanNodes` could pass as function args, then, we only need create one array list, not many.

File: be/src/exec/pipeline/pipeline_driver.h [hunk 8]
```diff
@@ -274,7 +274,21 @@ class PipelineDriver {
...
+    bool is_precondition_block() {
+        if (!_wait_global_rf_ready) {
+            if (dependencies_block() || local_rf_block()) {
+                return true;
+            }
+            _wait_global_rf_ready = true;
+            if (_global_rf_descriptors.empty()) {
+                return false;
+            }
+            _global_rf_wait_timeout_ns += _precondition_block_timer_sw->elapsed_time() + 5000000;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver.h:286]: Why not directly change `_global_rf_wait_timeout_ns`?

Comment (satanson) [reply be/src/exec/pipeline/pipeline_driver.h:286]: according to RuntimeFilterProbeCollector::wait(), actually, it wait for 25ms. so 5ms is added

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java (hunk 7 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:202]: ok

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 4 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:318]: ok

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java (hunk 5 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1326]: here, input_bytes is used to compute dop of the fragment instance,  input_bytes is more significant, for instance,  assuming a HashJoinScanNode follows the OlapScanNode, the larger avgRowSize, the more time to spend on HashJoinProbe computation.

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java (hunk 6 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:938]: these code is unused.
the evaluated row count of OlapScanNode is in-exact, so here just use numScanRanges and degreeOfParallelism to compute dop of ScanNode.

**Review by ZiheLiu — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by Pslydhh — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**